### PR TITLE
Detect/utilize consumers package manager of choice

### PR DESCRIPTION
**Pull request by hanford**
This PR attempts to determine what the users package manager of choice is, and uses it to install the `vercel` CLI

Closes https://github.com/cloudflare/next-on-pages/issues/6

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -569,3 +585,40 @@ const main = async ({
...
+      return "yarn";
+    } else if (userAgent.startsWith("pnpm")) {
+      return "pnpm";
+    } else {
+      return "npm";
+    }
+  }
+
+  const hasYarnLock = await exists("yarn.lock");
+  const hasPnpmLock = await exists("pnpm-lock.yaml");
```
Comment (hanford) [inline src/index.ts:615]: For [PNPM workspaces](https://pnpm.io/workspaces) it's possible we'd want to traverse up until we find a `pnpm-lock.yaml` which would be at the workspace root.

Alternatively package manager could be an option ... _or_ we could just add some documentation asking users to set `npm_config_user_agent`.

I'm in different / am happy to tweak this PR in any way

File: src/index.ts [hunk 2]
```diff
@@ -569,3 +585,40 @@ const main = async ({
...
+    if (userAgent.startsWith("yarn")) {
+      return "yarn";
+    } else if (userAgent.startsWith("pnpm")) {
+      return "pnpm";
+    } else {
+      return "npm";
+    }
+  }
+
+  const hasYarnLock = await exists("yarn.lock");
```
Comment (daniharo) [inline src/index.ts:614]: I'd do `if (hasYarnLock) return "yarn"` just after this line to avoid unnecessary checking for pnpm-lock.

**Comment by daniharo** [comment]
This line should say `npm i -D vercel`, `pnpm i -D vercel` or `yarn add -D vercel`, depending on detected package manager.

https://github.com/cloudflare/next-on-pages/blob/3db5504ec66f8166426efc6772d3a1abc9ab75dd/src/index.ts#L85

`npx vercel build` was already wrong as this was always `npm i -D vercel`

File: src/index.ts [hunk 3]
```diff
@@ -38,7 +38,23 @@ const prepVercel = async () => {
...
+      console.log("⚡️");
+      vercelBuild = spawn("pnpm", ["add", "-D", "vercel"]);
+      break;
+    case "yarn":
+      console.log('⚡️ Detected "yarn", running yarn add -D vercel');
+      console.log("⚡️");
+      vercelBuild = spawn("yarn", ["add", "-D", "vercel"]);
+      break;
+    default:
+      vercelBuild = spawn("pnpm", ["add", "-D", "vercel"]);
```
Comment (daniharo) [inline src/index.ts:56]: Default case would be unaltered as I understand.
```suggestion
      vercelBuild = spawn("npm", ["install", "-D", "vercel"]);
```

**Comment by hanford** [comment]
@daniharo not necessarily if you're using Direct Upload

**Comment by jokull** [comment]
Tested this locally - works for me!

**Comment by dario-piotrowicz** [comment]
@hanford thanks a lot for the PR (and sorry for it not having been reviewed in quite some time 🙇)

We've just merged https://github.com/cloudflare/next-on-pages/pull/71 as it looked like a simpler and more performant fix for the issue, sorry about it, if you can please check the latest commit on main and let us know if that works for you

I'm closing the PR since it's no longer needed/applicable, if you disagree feel free to reopen it

**Pull request closed without merging**